### PR TITLE
[BEAM-7933] Add job server request timeout (default to 60 seconds)

### PR DESCRIPTION
**Pull request by ecanzonieri**
Add a new option --job_server_timeout that default to 60 seconds. The request timeout is user for all job server requests (with exception to the ones that are expected to last long or hang). The request timeout is also used upon channel creation so that the Beam driver will fail if the job server is not available.

Let me know if the option name looks fine or it needs renaming. Finally, there are other grpc requests (non job service) that I'm not covering as part of this pr. Let me know if I should add the timeout to those as well.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by ecanzonieri** [comment]
R: @ibzib @aaltay

File: sdks/python/apache_beam/options/pipeline_options.py [hunk 1]
```diff
@@ -808,11 +808,15 @@ class PortableOptions(PipelineOptions):
...
-                        default=None,
-                        help=
-                        ('Job service endpoint to use. Should be in the form '
-                         'of address and port, e.g. localhost:3000'))
+    parser.add_argument(
+        '--job_endpoint', default=None,
+        help=('Job service endpoint to use. Should be in the form of address '
+              'and port, e.g. localhost:3000'))
+    parser.add_argument(
+            '--job_server_timeout', default=60, type=int,
```
Comment (ibzib) [inline sdks/python/apache_beam/options/pipeline_options.py]: indentation here looks off

File: sdks/python/apache_beam/runners/portability/portable_runner.py [hunk 2]
```diff
@@ -251,7 +253,11 @@ def send_options_request(max_retries=5):
           # This reports channel is READY but connections may fail
           # Seems to be only an issue on Mac with port forwardings
           return job_service.DescribePipelineOptions(
-              beam_job_api_pb2.DescribePipelineOptionsRequest())
+              beam_job_api_pb2.DescribePipelineOptionsRequest(),
+                  timeout=portable_options.job_server_timeout)
+        except grpc.FutureTimeoutError:
+          # no retry for timeout errors
```
Comment (ibzib) [inline sdks/python/apache_beam/runners/portability/portable_runner.py]: Are we sure we don't want retries? We could do exponential backoff, similar to what SubprocessJobServer already does.

File: sdks/python/apache_beam/runners/portability/portable_runner.py [hunk 3]
```diff
@@ -141,23 +141,25 @@ def _create_environment(options):
           payload=(portable_options.environment_config.encode('ascii')
                    if portable_options.environment_config else None))
 
-  def default_job_server(self, options):
+  def default_job_server(self, portable_options):
```
Comment (ibzib) [inline sdks/python/apache_beam/runners/portability/portable_runner.py]: why this change?

File: sdks/python/apache_beam/runners/portability/portable_runner.py [hunk 4]
```diff
@@ -141,23 +141,25 @@ def _create_environment(options):
...
     # TODO Provide a way to specify a container Docker URL
     # https://issues.apache.org/jira/browse/BEAM-6328
     if not self._dockerized_job_server:
       self._dockerized_job_server = job_server.StopOnExitJobServer(
           job_server.DockerizedJobServer())
     return self._dockerized_job_server
 
-  def create_job_service(self, options):
-    job_endpoint = options.view_as(PortableOptions).job_endpoint
+  def create_job_service(self, portable_options):
```
Comment (ibzib) [inline sdks/python/apache_beam/runners/portability/portable_runner.py]: I'd just leave this as `options`.

File: sdks/python/apache_beam/runners/portability/portable_runner.py [hunk 5]
```diff
@@ -240,7 +242,7 @@ def run_pipeline(self, pipeline, options):
             known_runner_urns=flink_known_urns,
             partial=True)
 
-    job_service = self.create_job_service(options)
+    job_service = self.create_job_service(portable_options)
```
Comment (ibzib) [inline sdks/python/apache_beam/runners/portability/portable_runner.py]: Not sure this is necessary?

File: sdks/python/apache_beam/runners/portability/portable_runner.py [hunk 6]
```diff
@@ -251,7 +253,11 @@ def send_options_request(max_retries=5):
           # This reports channel is READY but connections may fail
           # Seems to be only an issue on Mac with port forwardings
           return job_service.DescribePipelineOptions(
-              beam_job_api_pb2.DescribePipelineOptionsRequest())
+              beam_job_api_pb2.DescribePipelineOptionsRequest(),
+                  timeout=portable_options.job_server_timeout)
```
Comment (ibzib) [inline sdks/python/apache_beam/runners/portability/portable_runner.py]: I think this line is causing tests to fail. Also, we're already retrying this request. So I would either keep the existing logic or add the new timeout, but not both.

File: sdks/python/apache_beam/runners/portability/portable_runner.py (hunk 4 above)
Comment (ecanzonieri) [reply sdks/python/apache_beam/runners/portability/portable_runner.py]: I changed it because it is actually immediately converted back to portable_options. Basically the caller takes options and converts it to portable_options, then to the create_job_service method it passes options which is again converted to portable_options. It seems to be that we could pass portable_options in the first place and simplify the code a bit. However, I don't have a strong option, I'll revert back to options.

File: sdks/python/apache_beam/runners/portability/portable_runner.py (hunk 6 above)
Comment (ecanzonieri) [reply sdks/python/apache_beam/runners/portability/portable_runner.py]: My guess is that this is invoking the LocalJobServicer that doesn't have the timeout argument. I'll try to repro locally to verify. If that's the case I'd like to add the timeout argument to all LocalJobServicer methods.

File: sdks/python/apache_beam/runners/portability/portable_runner.py (hunk 2 above)
Comment (ecanzonieri) [reply sdks/python/apache_beam/runners/portability/portable_runner.py]: My guess is that we want to retry for other errors, but we should treat timeout as unretriable, so that we never wait for longer than the specified timeout (including retries). We could implement exponential backoff so that it will overall still wait for no longer than the timeout, but I'm not sure it's worth the additional complexity.

File: sdks/python/apache_beam/runners/portability/portable_runner.py (hunk 3 above)
Comment (ecanzonieri) [reply sdks/python/apache_beam/runners/portability/portable_runner.py]: I reverted it.

File: sdks/python/apache_beam/runners/portability/portable_runner.py (hunk 6 above)
Comment (ecanzonieri) [reply sdks/python/apache_beam/runners/portability/portable_runner.py]: I fixed the tests by adding the timeout to the stubs of the LocalJobServicer.

File: sdks/python/apache_beam/runners/portability/portable_runner.py (hunk 2 above)
Comment (ibzib) [reply sdks/python/apache_beam/runners/portability/portable_runner.py]: Okay.

File: sdks/python/apache_beam/runners/portability/portable_runner.py (hunk 6 above)
Comment (ibzib) [reply sdks/python/apache_beam/runners/portability/portable_runner.py]: Adding the timeout to the stubs fixes the tests, but are you sure adding the timeout like that is actually effective? In all the other places you changed, you put timeout in the requests, but here you put it in the service method, which I assumed was the problem. ie `DescribePipelineOptions(DescribePipelineOptionsRequest(), timeout=...)` should be  `DescribePipelineOptions(DescribePipelineOptionsRequest(timeout=...))`

File: sdks/python/apache_beam/runners/portability/local_job_service.py [hunk 7]
```diff
@@ -169,7 +169,7 @@ def GetMessageStream(self, request, context=None):
         resp = beam_job_api_pb2.JobMessagesResponse(message_response=msg)
       yield resp
 
-  def DescribePipelineOptions(self, request, context=None):
+  def DescribePipelineOptions(self, request, context=None, timeout=None):
```
Comment (ibzib) [inline sdks/python/apache_beam/runners/portability/local_job_service.py]: It doesn't look like the `timeout` argument is actually used -- see other comment below

File: sdks/python/apache_beam/runners/portability/portable_runner.py (hunk 6 above)
Comment (ecanzonieri) [reply sdks/python/apache_beam/runners/portability/portable_runner.py]: hmmm, I may be missing something but the timeout is never added to the request. My understanding is that timeout itself is not a request parameter but rather a grpc method call parameter. @ibzib can you point me to the code where you see me adding the timeout to the request? Perhaps, I accidentally added it to the wrong place. 
I'm no grpc expert, but I've ran some manual tests on this feature:
1) run python driver without job server running and default timeout -> The driver times out as expected after 60 seconds.
2) run python driver without job server running and change timeout -> The driver times out as expected after the specified timeout.
3) run python driver with job server running but code changed to slow down Prepare() -> The driver times out as expected.
4) run python driver with long running streaming job (Run() method) -> the driver program never times out (note this is the expected behavior since the driver is stuck waiting for completion.

Comment (ibzib) [reply sdks/python/apache_beam/runners/portability/portable_runner.py]: Oh, I read everything wrong. :sweat_smile: 

It'd be great to have real tests to verify all these timeouts, but I guess they would not make suitable unit tests, as they would all require at least 1 second to complete.

**Comment by ibzib** [comment]
Please fix lint errors.

**Comment by ecanzonieri** [comment]
Yeah, sorry about the lint errors. I still need to find a good configuration for PyCharm that actually works for Beam.

**Comment by ecanzonieri** [comment]
Apparently `:sdks:python:test-suites:tox:py2:docs` failed. I'm locally to check whether related to my pr.

**Comment by ibzib** [comment]
Docs failure is probably https://issues.apache.org/jira/browse/BEAM-8286

**Comment by ibzib** [comment]
Run Python PreCommit

File: sdks/python/apache_beam/options/pipeline_options.py [hunk 8]
```diff
@@ -818,11 +818,15 @@ class PortableOptions(PipelineOptions):
...
-                         'of address and port, e.g. localhost:3000'))
+    parser.add_argument(
+        '--job_endpoint', default=None,
+        help=('Job service endpoint to use. Should be in the form of address '
+              'and port, e.g. localhost:3000'))
+    parser.add_argument(
+        '--job-server-timeout', default=60, type=int,
+        help=('Job service request timeout in seconds. The timeout '
+              'determines the max time the driver program will wait to '
+              'get a response from the job server.'))
```
Comment (ibzib) [inline sdks/python/apache_beam/options/pipeline_options.py]: Might want to make a note here that timeouts do not apply to the actual pipeline itself.

**Comment by ecanzonieri** [comment]
I messed up the rebase. I'll fix it and repush.

**Comment by ecanzonieri** [comment]
Squashed everything to a single commit now.
